presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/AbstractRowEncoder.java

```diff
@@ -0,0 +1,218 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.kafka.encoder;
+
+import com.google.common.primitives.Shorts;
+import com.google.common.primitives.SignedBytes;
+import io.prestosql.spi.block.Block;
+import io.prestosql.spi.connector.ConnectorSession;
+import io.prestosql.spi.type.SqlDate;
+import io.prestosql.spi.type.SqlTime;
+import io.prestosql.spi.type.SqlTimeWithTimeZone;
+import io.prestosql.spi.type.SqlTimestamp;
+import io.prestosql.spi.type.SqlTimestampWithTimeZone;
+import io.prestosql.spi.type.TimestampType;
+import io.prestosql.spi.type.TimestampWithTimeZoneType;
+import io.prestosql.spi.type.Type;
+
+import java.nio.ByteBuffer;
+import java.util.List;
+
+import static io.prestosql.spi.type.BigintType.BIGINT;
+import static io.prestosql.spi.type.BooleanType.BOOLEAN;
+import static io.prestosql.spi.type.DateType.DATE;
+import static io.prestosql.spi.type.DoubleType.DOUBLE;
+import static io.prestosql.spi.type.IntegerType.INTEGER;
+import static io.prestosql.spi.type.RealType.REAL;
+import static io.prestosql.spi.type.SmallintType.SMALLINT;
+import static io.prestosql.spi.type.TimeType.TIME;
+import static io.prestosql.spi.type.TimeWithTimeZoneType.TIME_WITH_TIME_ZONE;
+import static io.prestosql.spi.type.TinyintType.TINYINT;
+import static io.prestosql.spi.type.VarbinaryType.isVarbinaryType;
+import static io.prestosql.spi.type.Varchars.isVarcharType;
+import static java.lang.Float.intBitsToFloat;
+import static java.lang.Math.toIntExact;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public abstract class AbstractRowEncoder
+        implements RowEncoder
+{
+    protected final ConnectorSession session;
+    protected final List<EncoderColumnHandle> columnHandles;
+
+    /**
+     * The current column index for appending values to the row encoder.
+     * Gets incremented by appendColumnValue and set back to zero when the encoder is reset.
+     */
+    protected int currentColumnIndex;
+
+    protected AbstractRowEncoder(ConnectorSession session, List<EncoderColumnHandle> columnHandles)
+    {
+        this.session = requireNonNull(session, "session is null");
+        requireNonNull(columnHandles, "columnHandles is null");
+        this.columnHandles = validateColumns(columnHandles);
+        this.currentColumnIndex = 0;
+    }
+
+    /**
+     * Performs any checks on the column handle field values
+     */
+    protected abstract List<EncoderColumnHandle> validateColumns(List<EncoderColumnHandle> columnHandles);
```

Comment:
It cannot return different list than one we pass to it as an argument, right? Otherwise the `columHandles` stored here would not be compatible with loop which calls to `appendColumnValue`. Change to void (unless I miss something).

Reply:
It only returns a different list as a defensive copy, I don't think that this should affect how it functions

Reply:
I think we should do the defensive copy in `AbstractRowEncoder` and make `validate` return void so it is obvious it cannot change the list.

Reply:
right, would be best to have `validateColumns` only validate the columns rather than validate-and-copy the columns